src/main/host/descriptor/socket/inet/tcp.rs

```diff
@@ -60,54 +101,226 @@ impl TcpSocket {
         self.has_open_file = val;
     }
 
+    /// Update the current tcp state. The tcp state should only ever be updated through this method.
+    fn with_tcp_state<T>(
+        &mut self,
+        cb_queue: &mut CallbackQueue,
+        f: impl FnOnce(&mut tcp::TcpState<TcpDeps>) -> T,
+    ) -> T {
+        let rv = f(&mut self.tcp_state);
+
+        // we may have mutated the tcp state, so update the socket's file state and notify listeners
+
+        // if there are packets to send, notify the host
+        if self.tcp_state.wants_to_send() {
+            if let Some(socket) = self.socket_weak.upgrade() {
```

Comment:
Maybe worth a comment about why this might fail and/or a log message?

Reply:
Added.

Reply:
Changed this to a panic instead and added a comment.